testing/src/scenario/_runtime.py

```diff
@@ -224,7 +224,9 @@ def _virtual_charm_root(self):
         actions_yaml = virtual_charm_root / "actions.yaml"
 
         metadata_files_present: Dict[Path, Optional[str]] = {
-            file: file.read_text() if file.exists() else None
+            file: file.read_text()
+            if charm_virtual_root_is_custom and file.exists()
+            else None
```

Comment:
Saving the calls, it's this bit in the PR description:

> There's one additional performance tweak in the branch that doesn't make a significant difference, but is trivial to do: when checking if the YAML files exist, skip the filesystem exists() call if we just created an empty temporary directory a few lines above, since we know that it will never exist.

It's not going to make any real performance difference, but I feel it's generally cleaner and better to be checking a bool before you do external things (and even really before calling a function), if you're joining with `and`.